src/dmd/expressionsem.d

```diff
@@ -9636,36 +9636,115 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
-        if (tb.ty == Tsarray)
-            exp.type = tb.nextOf().arrayOf();
-        if (exp.type.ty == Tarray && tb1next && tb2next && tb1next.mod != tb2next.mod)
+        // new scope needed before of `goto Lrewrite`
         {
-            exp.type = exp.type.nextOf().toHeadMutable().arrayOf();
+            Type tb = exp.type.toBasetype();
+            if (tb.ty == Tsarray)
+                exp.type = tb.nextOf().arrayOf();
+            if (exp.type.ty == Tarray && tb1next && tb2next && tb1next.mod != tb2next.mod)
+            {
+                exp.type = exp.type.nextOf().toHeadMutable().arrayOf();
+            }
+            if (Type tbn = tb.nextOf())
+            {
+                if (exp.checkPostblit(sc, tbn))
+                    return setError();
+            }
+
+            Type t1 = exp.e1.type.toBasetype();
+            Type t2 = exp.e2.type.toBasetype();
+            if ((t1.ty == Tarray || t1.ty == Tsarray) &&
+                (t2.ty == Tarray || t2.ty == Tsarray))
+            {
+                // Normalize to ArrayLiteralExp or StringExp as far as possible
+                e = exp.optimize(WANTvalue);
+            }
+            else
+            {
+                //printf("(%s) ~ (%s)\n", e1.toChars(), e2.toChars());
+                result = exp.incompatibleTypes();
+                return;
+            }
         }
-        if (Type tbn = tb.nextOf())
+
+    Lrewrite:
+        if (auto ce = e.isCatExp())
         {
-            if (exp.checkPostblit(sc, tbn))
+            // Debug printing
+            bool printit = ce.loc.filename && ce.loc.filename[0] != '.';
+            if (printit)
+                printf("[%s:%d] %s\n", ce.loc.filename, ce.loc.linnum, ce.toChars());
+
+            Identifier hook = global.params.tracegc ? Id._d_arraycatnTXTrace : Id._d_arraycatnTX;
+            if (!verifyHookExist(ce.loc, *sc, hook, "construct array with other array", Id.object))
                 return setError();
-        }
-        Type t1 = exp.e1.type.toBasetype();
-        Type t2 = exp.e2.type.toBasetype();
-        if ((t1.ty == Tarray || t1.ty == Tsarray) &&
-            (t2.ty == Tarray || t2.ty == Tsarray))
-        {
-            // Normalize to ArrayLiteralExp or StringExp as far as possible
-            e = exp.optimize(WANTvalue);
-        }
-        else
-        {
-            //printf("(%s) ~ (%s)\n", e1.toChars(), e2.toChars());
-            result = exp.incompatibleTypes();
-            return;
+
+            auto arrayItems = new Expressions();
+            bool appendPreviousExpression(Expression exp)
+            {
+                // if exp is a previous call to _d_arraycatnTX takes its elements and append them to arrayItems
+                while (exp.isCastExp())
+                    exp = exp.isCastExp().e1;
+
+                auto comma = exp.isCommaExp();
+                if (!comma)
+                    return false;
+                if (auto e2Call = comma.e2.isCallExp())
```

Comment:
Not technically necessary but I suggest that you added curly braces to this `if`-statement.